File: api/java/toolgood.textfilter.api/src/main/java/toolgood/textfilter/api/Grpcs/TextFilterGrpcProvider.java
package toolgood.textfilter.api.Grpcs;

import java.util.ArrayList;
import java.util.List;

import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import toolgood.textfilter.api.Datas.Requests.TextFilterRequest;
import toolgood.textfilter.api.Datas.Requests.TextReplaceRequest;
import toolgood.textfilter.api.Datas.Texts.TextFilterContactItem;
import toolgood.textfilter.api.Datas.Texts.TextFilterDetailItem;
import toolgood.textfilter.api.Datas.Texts.TextFilterResult;
import toolgood.textfilter.api.Datas.Texts.TextReplaceResult;
import toolgood.textfilter.api.GrpcBase.TextFilterGrpcGrpc;
import toolgood.textfilter.api.GrpcBase.TextFilter.TextFilterContactGrpcResult;
import toolgood.textfilter.api.GrpcBase.TextFilter.TextFilterDetailGrpcResult;
import toolgood.textfilter.api.GrpcBase.TextFilter.TextFindAllGrpcReply;
import toolgood.textfilter.api.GrpcBase.TextFilter.TextReplaceGrpcReply;
import toolgood.textfilter.api.GrpcBase.TextFilterGrpcGrpc.TextFilterGrpcBlockingStub;
import toolgood.textfilter.api.Interfaces.ITextFilterProvider;

public class TextFilterGrpcProvider implements ITextFilterProvider {
    private final TextFilterGrpcBlockingStub blockingStub;

    public TextFilterGrpcProvider(String host, int port) {
        this(ManagedChannelBuilder.forAddress(host, port).usePlaintext().build());
    }

    public TextFilterGrpcProvider(String url) {
        this(ManagedChannelBuilder.forTarget(url).usePlaintext().build());
    }

    private TextFilterGrpcProvider(ManagedChannel channel) {
        blockingStub = TextFilterGrpcGrpc.newBlockingStub(channel);
    }

    @Override
    public TextFilterResult TextFilter(TextFilterRequest request) {
        TextFindAllGrpcReply response = blockingStub.textFilter(request.ToGrpcRequest());
        return Create(response);
    }

    @Override
    public TextReplaceResult TextReplace(TextReplaceRequest request) {
        TextReplaceGrpcReply response = blockingStub.textReplace(request.ToGrpcRequest());
        return Create(response);
    }

    @Override
    public TextFilterResult HtmlFilter(TextFilterRequest request) {
        TextFindAllGrpcReply response = blockingStub.htmlFilter(request.ToGrpcRequest());
        return Create(response);
    }

    @Override
    public TextReplaceResult HtmlReplace(TextReplaceRequest request) {
        TextReplaceGrpcReply response = blockingStub.htmlReplace(request.ToGrpcRequest());
        return Create(response);
    }

    @Override
    public TextFilterResult JsonFilter(TextFilterRequest request) {
        TextFindAllGrpcReply response = blockingStub.jsonFilter(request.ToGrpcRequest());
        return Create(response);
    }

    @Override
    public TextReplaceResult JsonReplace(TextReplaceRequest request) {
        TextReplaceGrpcReply response = blockingStub.jsonReplace(request.ToGrpcRequest());
        return Create(response);
    }

    @Override
    public TextFilterResult MarkdownFilter(TextFilterRequest request) {
        TextFindAllGrpcReply response = blockingStub.markdownFilter(request.ToGrpcRequest());
        return Create(response);
    }

    @Override
    public TextReplaceResult MarkdownReplace(TextReplaceRequest request) {
        TextReplaceGrpcReply response = blockingStub.markdownReplace(request.ToGrpcRequest());
        return Create(response);
    }

    private TextFilterResult Create(TextFindAllGrpcReply response) {
        TextFilterResult result = new TextFilterResult();
        result.code = response.getCode();
        result.riskCode = response.getRiskCode();
        result.requestId = response.getRequestId();
        result.message = response.getMessage();
        result.riskLevel = response.getRiskLevel();
        result.sentimentScore = response.getSentimentScore();
        List<TextFilterDetailItem> details = new ArrayList<TextFilterDetailItem>();
        int length = response.getDetailsCount();
        for (int i = 0; i < length; i++) {
            TextFilterDetailItem item = new TextFilterDetailItem();
            TextFilterDetailGrpcResult r = response.getDetails(i);
            item.Position = r.getPosition();
            item.RiskCode = r.getRiskCode();
            item.RiskLevel = r.getRiskLevel();
            item.Text = r.getText();
            details.add(item);
        }
        result.details = details;

        List<TextFilterContactItem> contactItems = new ArrayList<TextFilterContactItem>();
        length = response.getContactsCount();
        for (int i = 0; i < length; i++) {
            TextFilterContactItem item = new TextFilterContactItem();
            TextFilterContactGrpcResult r = response.getContacts(i);
            item.contactType = r.getContactType();
            item.contactString = r.getContactString();
            item.position = r.getPosition();
            contactItems.add(item);
        }
        result.contacts = contactItems;
        return result;
    }

    private TextReplaceResult Create(TextReplaceGrpcReply response) {
        TextReplaceResult result = new TextReplaceResult();
        result.code = response.getCode();
        result.requestId = response.getRequestId();
        result.message = response.getMessage();
        result.riskLevel = response.getRiskLevel();

        List<TextFilterDetailItem> details = new ArrayList<TextFilterDetailItem>();
        int length = response.getDetailsCount();
        for (int i = 0; i < length; i++) {
            TextFilterDetailItem item = new TextFilterDetailItem();
            TextFilterDetailGrpcResult r = response.getDetails(i);
            item.Position = r.getPosition();
            item.RiskCode = r.getRiskCode();
            item.RiskLevel = r.getRiskLevel();
            item.Text = r.getText();
            details.add(item);
        }
        result.details = details;

        return result;
    }

}

File: api/csharp/ToolGood.TextFilter.Api/Grpcs/TextFilterGrpcProvider.cs
#if GRPC
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using Grpc.Net.Client;
using ToolGood.TextFilter.Api.Datas.Requests;
using ToolGood.TextFilter.Api.Datas.Texts;
using ToolGood.TextFilter.Api.Interfaces;

namespace ToolGood.TextFilter.Api.Grpcs
{
    public class TextFilterGrpcProvider : ITextFilterProvider
    {
        private readonly TextFilterGrpc.TextFilterGrpcClient _client;

        public TextFilterGrpcProvider(string grpcHost)
        {
            var channel = GrpcChannel.ForAddress(grpcHost);
            _client = new TextFilterGrpc.TextFilterGrpcClient(channel);
        }

        public TextFilterResult HtmlFilter(TextFilterRequest request)
        {
            var reply = _client.HtmlFilter(request.ToGrpcRequest());
            return Create(reply);
        }

        public async Task<TextFilterResult> HtmlFilterAsync(TextFilterRequest request)
        {
            var reply = await _client.HtmlFilterAsync(request.ToGrpcRequest());
            return Create(reply);
        }

        public TextReplaceResult HtmlReplace(TextReplaceRequest request)
        {
            var reply = _client.HtmlReplace(request.ToGrpcRequest());
            return Create(reply);
        }

        public async Task<TextReplaceResult> HtmlReplaceAsync(TextReplaceRequest request)
        {
            var reply = await _client.HtmlReplaceAsync(request.ToGrpcRequest());
            return Create(reply);
        }

        public TextFilterResult JsonFilter(TextFilterRequest request)
        {
            var reply = _client.JsonFilter(request.ToGrpcRequest());
            return Create(reply);
        }

        public async Task<TextFilterResult> JsonFilterAsync(TextFilterRequest request)
        {
            var reply = await _client.JsonFilterAsync(request.ToGrpcRequest());
            return Create(reply);
        }

        public TextReplaceResult JsonReplace(TextReplaceRequest request)
        {
            var reply = _client.JsonReplace(request.ToGrpcRequest());
            return Create(reply);
        }


        public async Task<TextReplaceResult> JsonReplaceAsync(TextReplaceRequest request)
        {
            var reply = await _client.JsonReplaceAsync(request.ToGrpcRequest());
            return Create(reply);
        }

        public TextFilterResult MarkdownFilter(TextFilterRequest request)
        {
            var reply = _client.MarkdownFilter(request.ToGrpcRequest());
            return Create(reply);
        }
        public async Task<TextFilterResult> MarkdownFilterAsync(TextFilterRequest request)
        {
            var reply = await _client.MarkdownFilterAsync(request.ToGrpcRequest());
            return Create(reply);
        }

        public TextReplaceResult MarkdownReplace(TextReplaceRequest request)
        {
            var reply = _client.MarkdownReplace(request.ToGrpcRequest());
            return Create(reply);
        }

        public async Task<TextReplaceResult> MarkdownReplaceAsync(TextReplaceRequest request)
        {
            var reply = await _client.MarkdownReplaceAsync(request.ToGrpcRequest());
            return Create(reply);
        }
        public TextFilterResult TextFilter(TextFilterRequest request)
        {
            var reply = _client.TextFilter(request.ToGrpcRequest());
            return Create(reply);
        }

        public async Task<TextFilterResult> TextFilterAsync(TextFilterRequest request)
        {
            var reply = await _client.TextFilterAsync(request.ToGrpcRequest());
            return Create(reply);
        }
        public TextReplaceResult TextReplace(TextReplaceRequest request)
        {
            var reply = _client.MarkdownReplace(request.ToGrpcRequest());
            return Create(reply);

        }

        public async Task<TextReplaceResult> TextReplaceAsync(TextReplaceRequest request)
        {
            var reply = await _client.TextReplaceAsync(request.ToGrpcRequest());
            return Create(reply);
        }

        private TextReplaceResult Create(TextReplaceGrpcReply reply)
        {
            TextReplaceResult result = new TextReplaceResult() {
                Code = reply.Code,
                RiskLevel = reply.RiskLevel,
                Message = reply.Message,
                RequestId = reply.RequestId,
                ResultText = reply.ResultText
            };
            result.Details = new List<TextFilterDetailItem>();
            foreach (var detail in reply.Details) {
                TextFilterDetailItem item = new TextFilterDetailItem() {
                    RiskCode = detail.RiskCode,
                    RiskLevel = detail.RiskLevel,
                    Position = detail.Position,
                    Text = detail.Text,
                };
                result.Details.Add(item);
            }
            return result;
        }

        private TextFilterResult Create(TextFindAllGrpcReply reply)
        {
            TextFilterResult result = new TextFilterResult() {
                Code = reply.Code,
                RiskCode = reply.RiskCode,
                RiskLevel = reply.RiskLevel,
                Message = reply.Message,
                RequestId = reply.RequestId,
                SentimentScore = reply.SentimentScore
            };
            result.Details = new List<TextFilterDetailItem>();
            foreach (var detail in reply.Details) {
                TextFilterDetailItem item = new TextFilterDetailItem() {
                    RiskCode = detail.RiskCode,
                    RiskLevel = detail.RiskLevel,
                    Position = detail.Position,
                    Text = detail.Text,
                };
                result.Details.Add(item);
            }
            result.Contacts = new List<TextFilterContactItem>();
            foreach (var contact in reply.Contacts) {
                TextFilterContactItem item = new TextFilterContactItem() {
                    ContactType = contact.ContactType,
                    ContactString = contact.ContactString,
                    Position = contact.Position
                };
                result.Contacts.Add(item);
            }
            return result;
        }

    }
}

#endif